aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -799,4 +799,294 @@ Tensor sspaddmm(const Tensor& self, const Tensor& mat1, const Tensor& mat2,
+Tensor _sparse_sum(const SparseTensor& input) {
+  return input.coalesce().values().sum();
+}
+
+Tensor _sparse_sum(const SparseTensor& input, ScalarType dtype) {
+  // don't have to do a conversion to the correct dtype first
+  // just need to setup the accumulator correctly
+  return input.coalesce().values().sum(dtype);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum) {
+  return at::_sparse_sum(input, dims_to_sum, false);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, ScalarType dtype) {
+  return at::_sparse_sum(input.to(dtype), dims_to_sum, false);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, bool keepdim, ScalarType dtype) {
+  return at::_sparse_sum(input.to(dtype), dims_to_sum, keepdim);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, bool keepdim) {
+  AT_CHECK(input.is_coalesced(), "sparse_sum: sparse_sum doesn't support uncoalesced SparseTensor");
+
+  const int64_t input_dim = input.dim();
+  auto dims_to_sum_b = dim_list_to_bitset(dims_to_sum, input_dim);
+  auto dims_to_sum_v = dims_to_sum.vec();
+  maybe_wrap_dims(dims_to_sum_v, input_dim);
+
+  LongTensor indices = input._indices();
+  Tensor values = input._values();
+  IntList sizes = input.sizes();
+  const int64_t sparse_dim = input.sparse_dim();
+  const int64_t dense_dim = input.dense_dim();
+
+  auto dims_to_keep_v = std::vector<int64_t>();
+  auto dense_dims_to_sum_v = std::vector<int64_t>();
+  for (int64_t d = 0; d < input_dim; d++) {
+    if (dims_to_sum_b[d]) {
+      if (d >= sparse_dim) dense_dims_to_sum_v.emplace_back(d + 1 - sparse_dim);
+    }
+    else {
+      dims_to_keep_v.emplace_back(d);
+    }
+  }
+  const int64_t sparse_dims_to_sum_size = dims_to_sum_v.size() - dense_dims_to_sum_v.size();
+  const bool sum_all_sparse_dim = (sparse_dim == sparse_dims_to_sum_size);
+  const bool sum_dense_dim = (dense_dims_to_sum_v.size() > 0);
+
+  // new values
+  Tensor new_values;
+  if (sum_dense_dim) {
+    new_values = values.sum(dense_dims_to_sum_v);
+    if (keepdim) {
+      auto new_values_size = values.sizes().vec();
+      for (auto d : dense_dims_to_sum_v) new_values_size[d] = 1;
+      new_values = new_values.view(new_values_size);
+    }
+  }
+  else {
+    new_values = values.clone();
+  }
+
+  if (sum_all_sparse_dim) {
+    // return a dense tensor if sum over all sparse dims
+    new_values = new_values.sum(0);
+    if (keepdim) {
+      for (int64_t i = 0; i < sparse_dim; i++) {
+        new_values = new_values.unsqueeze(0);
+      }
+    }
+    return new_values;
+  }
+  else {
+    // new indices
+    LongTensor new_indices;
+    if (sparse_dims_to_sum_size == 0) {
+      new_indices = indices.clone();
+    }
+    else {
+      if (keepdim) {
+        new_indices = at::zeros_like(indices);
+        if (!sum_all_sparse_dim) {
+          for (int64_t d = 0; d < sparse_dim; d++) {
+            if (!dims_to_sum_b[d]) new_indices[d].copy_(indices[d]);
```

Comment:
keepdim on a sparse dim seems weird. Does this mean that we'll have `0` indices for sparse dims that asked to be summed up? What does that even mean? does it mean user has to call `coalesce()` for final result?

Reply:
Yes, when keepdim=True, there will be all zeros at the summed rows of indices tensor, and then we will call `coalesce()` for the user. But after that, there will still have rows with all zeros at indices tensor. Not sure how useful is keepdim for sparse users, we can remove it if the returned indices tensor causes too much surprise.